src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -293,115 +376,306 @@ public bool TryGetValue(string key, out object value)
+                return false;
+            }
+
+            public override bool ContainsKey(string key)
+            {
+                for (var i = 0; i < _properties.Length; i++)
                 {
-                    _index++;
-                    if (_index < _dictionary.Properties.Length)
+                    var property = _properties[i];
+                    if (string.Equals(key, property.Name, StringComparison.OrdinalIgnoreCase))
                     {
-                        var property = _dictionary.Properties[_index];
-                        var value = property.ValueGetter(_dictionary.Value);
-                        Current = new KeyValuePair<string, object>(property.Name, value);
                         return true;
                     }
                 }
 
-                Current = default(KeyValuePair<string, object>);
                 return false;
             }
 
-            public void Reset()
+            public override bool TrySetValue(string key, object value)
             {
-                Current = default(KeyValuePair<string, object>);
-                _index = -1;
-                _enumerator = _dictionary?.InnerDictionary == null ?
-                    default(Dictionary<string, object>.Enumerator) :
-                    _dictionary.InnerDictionary.GetEnumerator();
+                // PropertyStorage never sets a value.
+                return false;
+            }
+
+            public override void Upgrade(ref Storage storage)
+            {
+                storage = new ListStorage(Count);
+                for (var i = 0; i < _properties.Length; i++)
+                {
+                    var property = _properties[i];
+                    storage.TrySetValue(property.Name, property.GetValue(_value));
+                }
+            }
+
+            private static void ValidatePropertyNames(Type type, PropertyHelper[] properties)
+            {
+                var names = new Dictionary<string, PropertyHelper>(StringComparer.OrdinalIgnoreCase);
+                for (var i = 0; i < properties.Length; i++)
+                {
+                    var property = properties[i];
+
+                    PropertyHelper duplicate;
+                    if (names.TryGetValue(property.Name, out duplicate))
+                    {
+                        var message = Resources.FormatRouteValueDictionary_DuplicatePropertyName(
+                            type.FullName,
+                            property.Name,
+                            duplicate.Name,
+                            nameof(RouteValueDictionary));
+                        throw new InvalidOperationException(message);
+                    }
+
+                    names.Add(property.Name, property);
+                }
+            }
+        }
+
+        internal class EmptyStorage : Storage
+        {
+            public static readonly EmptyStorage Instance = new EmptyStorage();
+
+            public override int Count => 0;
```

Comment:
Suggest making the parameterless constructor `private` to keep future-Ryans honest.